### ISSUE TITLE
Image based Point Cloud Cropping in LiDAR Based Implementation 

### ISSUE THREAD

**Issue by sakshi404** [qianguih/voxelnet#9]
Hello,

Since VoxelNet deals with only LiDAR data to detect objects, Why cropping of data based on projection on image plane has been done in this implementation?
If this is done to make Point Cloud data sparse, it can be done through random sampling as mentioned in Research paper on VoxelNet.
Can you please comment on the issues faced while using jaesinema's work https://github.com/jeasinema/VoxelNet-tensorflow.

Thanks,
Sakshi

**Comment by johleh** [comment]
the reason for cropping is that the KITTI dataset has labels only for objects visible in the image i think.

**Comment by qianguih** [comment]
This is a common strategy used in literature. And I agree with @johleh.

**Comment by sakshi404** [comment]
Then how to detect objects solely on the basis of LiDAR data rather than
using image data for reference.

**Comment by sakshi404** [comment]
@qianguih  Can you please comment  on above query?
Is there any dataset which deals with only LiDAR data and doesn't use images for reference?
Have you worked on this type of problem?

**Comment by johleh** [comment]
during training cropping is used to avoid that good examples which are not visible in the image become negative examples because of missing labels. (that is often the case for parking cars the kitti car passes)
during inference it is better to crop the point cloud to not submit proposals in a space that is not visible in the image, and therefore has no labels as well. or to later delete proposal boxes that ly outside the visible area.

a quick and dirty way to do the cropping without using image data is to just cut the point cloud with 2 planes to simulate the image viewing frustum and delete everything that is outside of them.

image information is used only because of the way the dataset is labled, which makes cropping beneficial. no image information is used to generate predictions.

**Comment by sakshi404** [comment]
You mean to say that Kitti dataset has no labels for LiDAR data and has
labels only for image, so we need to crop whole LiDAR dataset as per the
images. But in kitti dataset, we already have LiDAR data corresponding to
images, so instead of doing calibration, we can directly use the LiDAR data
for training and testing. Don't you think we should use raw data for
testing even if we have to use cropped LiDAR data for training.

**Comment by johleh** [comment]
kitti labels are in camera coordinate space .

label2 for frame 1:
Truck 0.00 0 -1.57 599.41 156.40 629.75 189.25 2.85 2.63 12.34 0.47 1.49 69.44 -1.56
Car 0.00 0 1.85 387.63 181.54 423.81 203.12 1.67 1.87 3.69 -16.53 2.39 58.49 1.57
Cyclist 0.00 3 -1.65 676.60 163.95 688.98 193.93 1.86 0.60 2.02 4.59 1.32 45.84 -1.55
DontCare -1 -1 -10 503.89 169.71 590.61 190.13 -1 -1 -1 -1000 -1000 -1000 -10   // just the 2D box in the image)
...

Car 0.00 0 1.85 387.63 181.54 423.81 203.12 1.67 1.87 3.69 -16.53 2.39 58.49 1.57
2D image box : 387.63 181.54 423.81 203.12
3D box : 1.67 1.87 3.69 -16.53 2.39 58.49 1.57  (in camera coordinates)

calibration is used to shift the center of the label, because the lidar system and the left camera are located in different places atop the kitti car. look at the kitti documentation for further information.

you need to apply the calibration matrices on your predictions, if you want to use eval scripts, because they use labels in camera coordinates/position

you can run testing without cropping the lidar data, but you will get predictions for objects that are not visible in the image (left and right corners of the bird's eye view width x depth) and will be evaluated to false positives, because there are no labels. even though your predictions may be correct.

**Comment by sakshi404** [comment]
If I perform object detection using camera and LiDAR data separately
without doing any crop operation and then apply calibration to the
predictions, will I get correct results? My main concern is whether its
possible to perform object detection solely on the basis of LiDAR data.
Actually by cropping everything with reference to image, we are relying
heavily on Image Data, but I want to give equal weight to LiDAR and Camera
data.

**Comment by johleh** [comment]
voxelnet is solely based on lidar data, because it is supervised learning, labels are needed for training and evaluation.

calibration is needed and does no rely on image data, but on the used sensors and their position atop the car.
cropping does not rely on image data, using images is just a comfortable way to do it, but cutting the corners of the point cloud would not be that far off i think.
cropping is needed because for training only those regions of the point cloud are useful, that are actually labeled. same goes for predicting objects only in regions that are labeled to evaluate the model.

if all objects in the 360° pountcloud would be labeled, no cropping would be needed.
if labels for training/evaluation would be based on lidar coordinates, calibration would not be needed.

if you combine  voxelnet and image data, something similar is used in [this paper]( https://arxiv.org/abs/1711.06703) .

**Comment by sakshi404** [comment]
cropping does not rely on image data, using images is just a comfortable
way to do it, but cutting the corners of the point cloud would not be that
far off i think.
>>
If I don't want to use image data for reference, what can be the other ways
to do cropping ? Also, cropping is required only for training and
evaluation. right? During testing, algorithm will automatically ignore the
redundant data of LiDAR as done in image classification algorithms.

If I get a dataset with separate labels for image and LiDAR, I don't need
to do any calibration in  training and evaluation and only have to only
worry about labels in testing.

**Comment by sakshi404** [comment]
@johleh @qianguih 
Can you comment on the use of Deconv layer in this approach. How deconv layer is actually implemented. Some articles mentions the use of gradient and some are focused on just changing the weights and  input patterns and thus performing convolution. Please comment if are familiar with it

**Comment by sakshi404** [comment]
@johleh @qianguih  Does this model provide results for test data as well ? Can you tell me how it works on test dataset in term of functions implemented? As we discussed above, we use image data during training and validation for calibration, image data should not be used during testing.

I can see we are calling predict_step function for both training and evaluation. I can't see how are you calculating accuracy for validation set and how can you get results for testing data?

**Comment by johleh** [comment]
kittti provides an eval script which needs the labels predicted by your model.
or use the one provided here : https://github.com/traveller59/kitti-object-eval-python
to create the labels look at voxelnet/utils/utils.py. "box3d_to_label".

test works exactly like validation, you can use the calibration matrices provided for the test set to cut the pointclouds. then you have to submit these labels to the kitti server, because there are no ground truth labels available.

**Comment by sakshi404** [comment]
@johleh  @qianguih  in the pre-processing module, you are defining voxel size, grid size according to class category (car or something else). How these values will be defined for test data as we wont be having idea whether its car/pedestrian or something else?? For training, it is fine to define parameters like that.
One more thing, since we are associating the point cloud data to voxels, what kind of clustering it can be called? I was not getting how this clustering /voxel association is done in code.

**Comment by johleh** [comment]
i am not sure i understand your question. you train the model with a certain detection range and voxelsize. 70.4 x 80 x 4 meters with 0.2 x 0.2 x 0.4 meters sizes => 352 x 400 x 10 voxelgrid
you can then apply your model on your test lidar points within the detection range. you do not change this. the trained model takes the pointcloud and applies the voxelgrid to the testdata to calculate per-voxel features, which are then fed into the convolutional pipeline to generate predictions.

car/pedestrian/cyclist are individual models in the paper, trained on different detection ranges. (smaller detection range - higher resolution for pedestrians)
if you train them all together you have to solve additional problems first and make tradeoffs.

**Comment by sakshi404** [comment]
@johleh  Since we are converting point clouds to voxels by associating them to voxels. On what basis we can discard one point over other. What if we loose useful information due to this. Suppose we have 10 ponts in a voxel and maximum point sin voxel can be 7 then which three points will  be discarded.
Can we associate the voxel partition process to some kind of clustering algorithm?
Can you share your email id.

**Comment by johleh** [comment]
sampling points should not be a problem. those voxels that contain more then 35 points  are usually very near and should belong mostly to objects that are easy to detect (if not occluded).

you can look at a histogramm of points/number of voxels and then decide if it is important.
like 99.5 to 99.8% of the voxels are empty, and those are not empty usually contain only a few points.

**Comment by sakshi404** [comment]
@johleh  I wanted to know whether we can relate it to some kind of clustering?

**Comment by johleh** [comment]
as it is the association between points is simply based on the voxel grid and the points locations.
using clustering to change that ? i don't know.

**Comment by sakshi404** [comment]
VoxelNet paper reported 225ms inference time, can you tell me what was the inference time for this implementation?
